game/scripts/vscripts/abilities/oaa_arcane_orb.lua

```diff
@@ -31,175 +43,209 @@ function modifier_oaa_arcane_orb:RemoveOnDeath()
+  local ability = self:GetAbility()
 
-  -- process_procs == true in OnAttack means this is an attack that attack modifiers should not apply to
-  if keys.attacker ~= parent or keys.process_procs then
+  if event.attacker ~= parent then
     return
   end
 
-  local ability = self:GetAbility()
-  local target = keys.target
-  -- Wrap in function to defer evaluation
-  local function autocast()
-    return (
-      target.GetUnitName and -- Check for existence of GetUnitName method to determine if target is a unit
-      ability:GetAutoCastState() and
-      not parent:IsSilenced() and
-      ability:IsOwnersManaEnough() and
-      ability:IsOwnersGoldEnough(parent:GetPlayerOwnerID()) and
-      ability:IsCooldownReady() and
-      ability:CastFilterResultTarget(target) == UF_SUCCESS
-    )
+  if parent:IsIllusion() then
+	return
   end
 
-  if parent:GetCurrentActiveAbility() ~= ability and not autocast() then
+  -- To prevent crashes:
+  local target
+  if event.target == nil then
     return
+  else
+    target = event.target
   end
 
-  parent:RemoveModifierByName("modifier_oaa_arcane_orb_sound")
-  parent:ChangeAttackProjectile()
+  if target:IsNull() then
+    return
+  end
 
-  ability:CastAbility()
-  -- Enable proc for this attack record number
-  self.procRecords[keys.record] = true
+  if ability:IsOwnersManaEnough() and ability:IsCooldownReady() and (not parent:IsSilenced()) and (not target:IsMagicImmune()) then
+    if ability:GetAutoCastState() == true then
+      --The Attack while Autocast is ON
+      parent:AddNewModifier(parent, ability, "modifier_oaa_arcane_orb_sound", {})
+	  -- Change Attack Projectile
+      parent:SetRangedProjectileName("particles/units/heroes/hero_obsidian_destroyer/obsidian_destroyer_arcane_orb.vpcf")
+    else
+      --The Attack while Autocast is OFF
+      if parent:GetCurrentActiveAbility() == ability then
+        -- Arcane Orb Manual Cast
+        parent:AddNewModifier(parent, ability, "modifier_oaa_arcane_orb_sound", {})
+        parent:SetRangedProjectileName("particles/units/heroes/hero_obsidian_destroyer/obsidian_destroyer_arcane_orb.vpcf")
+      end
+    end
+  end
 end
 
-function modifier_oaa_arcane_orb:OnAttackLanded(keys)
+function modifier_oaa_arcane_orb:OnAttackLanded(event)
   local parent = self:GetParent()
-  local attackRecord = keys.record
-  local target = keys.target
   local ability = self:GetAbility()
+  local target = event.target
+
+  if event.attacker ~= parent then
+    return
+  end
+
+  -- To prevent crashes:
+  if target == nil then
+    return
+  end
 
-  if keys.attacker ~= parent or not self.procRecords[attackRecord] or not keys.process_procs then
+  if target:IsNull() then
     return
   end
 
-  local bonusDamage = parent:GetMana() * ability:GetSpecialValueFor("mana_pool_damage_pct") / 100
-  self.procRecords[attackRecord] = nil
+  if ability:IsOwnersManaEnough() and ability:IsCooldownReady() and (not parent:IsSilenced()) and (not target:IsMagicImmune()) then
```

Comment:
There's a reason for all the proc record stuff. This stuff shouldn't be getting checked here. What happens if the caster runs out of Mana while the attack is flying? Or if autocast is turned on while a normal attack is flying? Only magic immunity is properly relevant.

Reply:
Mana is removed **when the projectile lands**. Similar behavior to one before patch 7.14.
If autocast is turned on while normal attack is flying, arcane orb will be activated, bonus damage will be dealt and mana will be spent.
If caster runs out of Mana while the attack is flying, arcane orb will not do bonus damage, because IsOwnersManaEnough() will be false.

Reply:
I know about the mana being removed when it lands. I thought it was dumb behaviour when I found out, and so didn't bother matching it.
Does vanilla seriously just do all cast checks when it lands though? As in proccing when you toggle autocast while in flight? That's super jank.

Reply:
After patch 7.14. they made it so mana is removed after launching, instead of after landing. And that was a nerf to OD. Right now I did not bother with vanilla matching, I wanted to make it simple and working. If current behavior (with my implementation) is irritating you that match, I will try to match vanilla. But imo ''what is not broken should not be fixed''.